miditext.1
.TH MIDITEXT 1 2026-07-09 "Linux" "User Commands"
.SH NAME
miditext \- convert a MIDI 1.0 standard file to textmidi language
.SH SYNOPSIS
.SY miditext
.OP OPTION
.OP \-\-midi INPUT
.OP \-\-textmidi OUTPUT
.YS

.SH DESCRIPTION
.B miditext
Converts a text file description of a MIDI file to textmidi language
.B miditext 
is fully documented in a PDF file miditext.pdf.
.SH OPTIONS
.TP
.BI \-\-midi\  input_file
The input file; a standard MIDI 1.0 file.
.
.
.TP
.BI \-\-textmidi\  output_file
The output file; a text file in textmidi language.
.
.
.TP
.BI \-\-dotted_rhythms\  true|false
If true and lazy is selected, then dotted rhythms are written for the appropriate rhythms.
If false and lazy is selected, then rhythms are written without using dots.
.
.
.TP
.BI \-\-rhythmexpression\  rational|simplecontinuedfraction
The output file; a text file in textmidi language.
.
.
.TP
.BI \-\-dynamics_configuration\  {dynamics_file}
Specify the optional file of dynamic values for the lazy mode dynamics symbols, for example, mf, forte, fff etc.
The format of the file is of symbols with equal signs and MIDI velocity values for each symbol.
The symbols may be in any order.  It is not necessary to set all of the available symbols.
If a symbol does not appear, then a default value is used.
Format:
.nf
ppppp=5
pppp=10
ppp=25
pp=40
p=50
mp=62
mf=75
forte=90
ff=110
fff=120
ffff=127
fffff=127
.fi
.
.
.TP
.BI \-\-quantize\  \(aqnumerator/denominator\(aq
A musical rhythm ratio to which to quantize the durations.  
Quotes are typically needed in command shells.
For example "1/16" would quantize to a sixteenth note.
.
.
.TP
.B \-\-lazy
Generate output in textmidi "lazy" mode, an abbreviated
representation that represents musical notes rather than
note-on and note-off pairs.
.
.
.TP
.B \-\-answer
Ask to verify to overwrite the output file if the output file exists.
.
.
.TP
.B \-\-version
Print the version of the program.
.
.
.TP
.B \-\-verbose
Print additional information during processing.
.
.
.TP
.B \-\-help
Print the command line options.
.
.
.SH EXIT STATUS
.TP
.B miditext
always returns an EXIT_SUCCESS.
.SH ERRORS
.
Errors reported by
.B miditext
include:
.
.RS
.RS
.TP
There is extra data after the tracks which will be ignored.
.
.TP
No Track header MTrk in track:
.
.TP
File too short for track the length found in the track:
.
.TP
Program options error:
.
.TP
File does not exist.
.
.TP
You must provide a MIDI input file
.
.TP
The text and MIDI filenames are the same!  You would have overwritten the MIDI file!; Must exit!
.
.TP
Open failed for file
.
.TP
IOS exceptions.
.
.TP
ERROR File too short for a MIDI Header
.
.TP
MIDI channel exceeded MIDI limit
.         
.TP
SMPTE Offset out of range
. 
.TP
MIDI Key Number exceeded MIDI limit
.         
.TP
MIDI velocity exceeded MIDI limit
.
.TP
MIDI Pitch Bend LSB exceeded MIDI limit
.
.TP
MIDI Pitch Bend MSB exceeded MIDI limit
.
.TP
Illegal 8-bit control value:
.
.TP
Illegal pan value fails range:
. 
.TP
Illegal 14-bit control value: 
.
.TP
Illegal 8-bit control value:
.
.TP
MIDI pressure exceeded MIDI limit
. 
.TP
no noteoff found, channel:
.                 
.TP
Unrecognized byte sequence:
.
.TP
key counter went negative on channel: too many note-offs\n";
.
.TP
You must provide a MIDI input file
.
.TP
You would have overwritten the source name!; Must exit!
.
.TP
open failed
.
.SH FILES
.TP
.I midi_input_file
is the output of 
.B textmidi
.
It is a standard MIDI 1.0 file which can be used with programs that play and edit 
standard MIDI 1.0 files.
.
.
.TP
.I output_file
is a text file in the MIDI language supported by 
.B textmidi.
.
.
.TP
.I dynamics.cfg
or other name contains dynamic symbols assigned to MIDI velocities.

.SH CONFORMING TO
International MIDI Association MIDI Specification 1.1.
.SH EXAMPLE
.SH NOTES
.SH BUGS
Output in lazy mode is not fully qualified and may fail for 
overlapping durations.
.SH AUTHORS
Written by Thomas E. Janzen in 1999 and refined until release.
.SH SEE ALSO
.TP
.B textmidi


NEWS
2021-08-12
    * Version 1.0
      This is the first public version of textmiditools
      including rewrites of AlgoRhythms as textmidicgm,
      smusmidi as smustextmidi, mididisasm as miditext,
      and textmidi.
    * Version 1.0.1
    * Version 1.0.2
    * Version 1.0.3
      Fix a length problem: using a generated form using
      textmidicgm --random, the rhythms were very unusual
      and reveal a time-counting issue in the note-event
      generating loop in Composer.cc.
    * Version 1.0.4
      Update documentation for texmidicgm.
      Random forms now have a set .125 probability of silence,
      for walking voices.
      Panning for random forms has been fixed so that
      if there is only one channel it won't land on the
      far right of the audio field.
    * Version 1.0.5
      Fix bug of random forms getting different program
      assignments on different voices that have the same
      MIDI channel.   There must be a (single-valued relation)
      function from MIDI channel to program.
    * Version 1.0.6
      Add textmidiformedit.py and its associated python files,
      to make it possible to graphically edit musical forms
      for use by textmidicgm to compose random music.
      Also added previously omitted B's from the chromatic
      scale in textmidicgm for making random forms.
    * Version 1.0.7
      Fix form plot's unside-down mean curves, and add a
      Postscript menu item.
      Fix textmidiformedit.py form plots which had upsidedown
      mean curves; add a save-postscript menu item.
    * Version 1.0.8

    * Version 1.0.9
      2022-01-07
      Add a keyboard window for scale specification to textmidiformedit.py.
      Change the default detache (separation between events) to 0 in textmidi.l.
    * Version 1.0.10
      2022-01-17
      Fix TextMIDITools version in sources. Fix spelling of Association.
    * Version 1.0.11
      textmidiformedit.py has been made more like a package
      that can be installed from github or pypi.org, and run:
      python3
            from textmidiformedit import *
            textmidiformedit()
            etc.
    * Version 1.0.12
      Fixes bugs in reading/writing form files in textmidiformedit.py,
      including for scale, length, min/max note length, pulse,
      melody probabilities and form name.
      Add trackscramble feature to textmidicgm with trackscrambleperiod,
      so change-up the voice priority periodically under the texture curve.
    * Version 1.0.13
      Fixes more bugs in reading/writing form files in textmidiformedit.py,
      including for scale, length, min/max note length, pulse,
      melody probabilities and form name. Fixed bad assignment of
      values to widget variables, which is wrong. Used get/set
      on widget variables. Fixed behaviour of
      voice->follow radio buttons. textmidiformedit is in much
      better condition. The number of remaining bugs est. to be < 1,000,000.
    * Version 1.0.14
      Fixes more bugs in reading/writing form files in textmidiformedit.py;
      added resize to form plot window.  Use GeneralMIDI names in program entry.
    * Version 1.0.15
      2022-02-07
      Fixes more bugs in reading/writing form files in textmidiformedit.py.
      Also improved reliability of VoiceWIndow by removing most
      widget variables (e.g., StringVar and IntVar) for Spinbox
      and Entry as theydid not behave well and copoied data
      into the wrong voice.
      Added an update_idletasks)( to VoiceWIndow that seems to
      help so that data in entries and spinboes are not copied
      to other voices.
    * Version 1.0.16
      2022-02-10
    * Version 1.0.17
      textmidiformedit.py seems to work wel enough now.
      2022-02-10
    * Version 1.0.18
      2022-02-11
      Update copyright to 2022. Update to version 1.0.18. Reflect
      changes to textmidiformedit in textmidiformedit.texi.
      Fix ambiguity about quotes in text events in textmidi.texi.
    * Version 1.0.19
      2022-02-14
      No longer crashes due to third-party class library when
      compiled with C++20.
    * Version 1.0.20
      2022-02-18
      Fixed RhythmRational print_rhythm to print 4 whole notes
      as 4/1, because 4 is a quarter note in textmidi language's
      lazy mode.  Fixed SmusTrackEvent.cc mishandling of whole notes.
    * Version 1.0.21
      2022-02-21
      Fix smustextmidi's mishandling of dynamics in which a
      velocity of 64 would be used instead of the volume in
      an SMUS volume event;
      also ignore volume events that repeat the same dynamic
      as is current. Since MIDI has no volume events but puts
      velocity in every note event, this does not affect the
      ultimate MIDI file version.
      Removed a duplicate typedef of MidiStreamVector from Midi.h
    * Version 1.0.22
      2022-03-06
      Disabled --lazy mode for miditext because it just doesn't work.
      Move io_bytes to a separate header.
      Remove extra data structures and unused objects.
      Clean up dependencies.
      Fix smustextmidi printing 1/1 for a whole note (although this is acceptable) and change to "1".
      Make Composer a class.
      Put more namespaces under textmidi namespace.
      clean up dependencies of includes and on classes.
    * Version 1.0.23
      2022-03-06
      Fixed smustextmidi bug. It produced a tied-in-and-out
      note as in -C4- if there had been a previous tied note
      of the same pitch tied-over multiple times because it
      would add each new occurance in the tied note to tied_vec_
      in SmusTrackEvents.cc.
      Removed a couple unused things. textmidi would convert
      this into excess numbers of note-offs. Detected using MuseScore.
    * Version 1.0.24
      2022-03-17
      Always write the MIDI file format (e.g., MULTITRACK)
      when writing textmidi text files.  The programs that
      write textmidi text files write more uniform code.

miditext.cc
//
// TextMIDITools Version 1.1.6
//
// Copyright © 2025 Thomas E. Janzen
// License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>
// This is free software: you are free to change and redistribute it.
// There is NO WARRANTY, to the extent permitted by law.
//
// Written as mididisasm in 2003 as a companion to textmidi.
//
// miditext is the main driver for miditext,
// which converts a binary MIDI standard file
// into a textmidi text file.
//
// 2020-03-16
#if HAVE_CONFIG_H
#  include <config.h>
#endif /* HAVE_CONFIG_H */
#include <endian.h>

#include <sys/types.h>
#include <sys/stat.h>

#include <unistd.h>
#include <fcntl.h>

#include <cstdint>
#include <cstdlib>
#include <cctype>

#include <algorithm>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <iterator>
#include <map>
#include <memory>
#include <ranges>
#include <set>
#include <string>
#include <thread>
#include <vector>

#include <boost/algorithm/string/case_conv.hpp> // to_upper
#include <boost/lexical_cast.hpp>
#include <boost/program_options.hpp>

#include "DynamicsOptions.h"
#include "MidiEvents.h"
#include "MidiMaps.h"
#include "Options.h"
#include "TextmidiUtils.h"
#include "rational_support.h"

using std::size_t, std::pair, std::vector, std::cin, std::cerr,
      std::string, std::ranges::copy, std::toupper;
using midi::MidiStreamRange;
using textmidi::rational::RhythmRational,
      textmidi::rational::PrintRhythmRational,
      textmidi::DelayEvents, textmidi::DelayEvent;
using boost::lexical_cast;
namespace
{
    using std::equal;
    using TrackPositionLength = pair<size_t, size_t>;

    template <typename R>
    void find_tracks(const R&& midi_view,
        vector<TrackPositionLength>& track_bounds, size_t expected_track_qty)
    {
        track_bounds.clear();
        track_bounds.reserve(140); // Saw a file with 140 tracks.
        size_t midi_ctr{sizeof(midi::MidiHeader)};
        while (midi_ctr < midi_view.size())
        {
            if (!equal(&midi_view[midi_ctr],
                &midi_view[midi_ctr + midi::MidiTrackChunkName.size()],
                &midi::MidiTrackChunkName[0]))
            {
                const auto track_num{track_bounds.size() + 1};
                if (track_num > expected_track_qty)
                {
                    cerr << "There is extra data after "
                        "the tracks which will be ignored.\n";
                }
                else
                {
                    const string errstr{
                        (string{"No Track header MTrk in track: "}
                        += track_num) += '\n'};
                    cerr << errstr;
                }
                return;
            }
            midi_ctr += midi::MidiTrackChunkName.size();
            int32_t track_len{};
            copy(&midi_view[midi_ctr], &midi_view[midi_ctr] + (sizeof track_len),
                textmidi::io_bytes(track_len));
            midi_ctr += sizeof(track_len);
            // swap the bytes of the track length
            track_len = htobe32(track_len);
            if (std::cmp_less(midi_view.size() - midi_ctr, track_len))
            {
                const string errstr{(string{
                    "File too short for track the length found in the track:"}
                    += lexical_cast<string>(track_bounds.size())) += '\n'};
                cerr << errstr;
                return;
            }
            track_bounds.emplace_back(midi_ctr, track_len);
            midi_ctr += track_len;
        }
    }

    class ConvertTrack
    {
      public:
        ConvertTrack(DelayEvents& message_pairs,
            uint32_t ticks_per_whole, RhythmRational quantum, MidiStreamRange stream_range)
          :
            message_pairs_{message_pairs},
            ticks_per_whole_{ticks_per_whole},
            quantum_{quantum},
            stream_range_(stream_range)
        {
        }
        void operator()()
        {
            using std::optional, std::tie;
            textmidi::MidiEventFactory midi_event_factory{ticks_per_whole_};
            int64_t ticks_accumulated{};

            do
            {
                optional<DelayEvent> de;
                tie(stream_range_, de) = (midi_event_factory(stream_range_, ticks_accumulated));
                if (de.has_value())
                {
                    message_pairs_.push_back(std::move(de.value()));
                }
            }
            while (!stream_range_.empty());
        }
     private:
        DelayEvents& message_pairs_;
        const uint32_t ticks_per_whole_;
        const RhythmRational quantum_;
        MidiStreamRange stream_range_;
    };
} // namespace

int main(int argc, char *argv[])
{
    using boost::to_upper;
    using std::cout, std::filesystem::exists, std::filesystem::file_size,
        std::ifstream, std::istringstream, std::make_unique,
        std::ofstream, std::ranges::count_if, std::set, std::setw;
    using textmidi::PrintLazyTrack, textmidi::rational::RhythmExpression;
    boost::program_options::options_description desc("Allowed options");
    desc.add_options()
        ((help_option.registered_name().c_str()), help_option.text().c_str())
        ((verbose_option.registered_name().c_str()),
            verbose_option.text().c_str())
        ((version_option.registered_name().c_str()),
            version_option.text().c_str())
        ((midi_in_option.registered_name().c_str()),
            boost::program_options::value<string>(),
            midi_in_option.text().c_str())
        ((answer_option.registered_name().c_str()),
            answer_option.text().c_str())
        ((textmidi_out_option.registered_name().c_str()),
            boost::program_options::value<string>(),
            textmidi_out_option.text().c_str())
        ((quantize_option.registered_name().c_str()),
            boost::program_options::value<string>(),
            quantize_option.text().c_str())
        ((lazy_option.registered_name().c_str()),
            lazy_option.text().c_str())
        ((dynamics_configuration_option.registered_name().c_str()),
            boost::program_options::value<string>(),
            dynamics_configuration_option.text().c_str())
        ((dotted_rhythms_option.registered_name().c_str()),
            boost::program_options::value<string>(),
            dotted_rhythms_option.text().c_str())
        ((rhythm_expression_option.registered_name().c_str()),
            boost::program_options::value<string>(),
            rhythm_expression_option.text().c_str())
    ;
    boost::program_options::positional_options_description pos_opts_desc;
    boost::program_options::variables_map var_map;
    try
    {
        pos_opts_desc.add(midi_in_option.option().c_str(), -1);
        boost::program_options::store(
            boost::program_options::command_line_parser(argc, argv)
            .options(desc).positional(pos_opts_desc).run(), var_map);
        boost::program_options::notify(var_map);
    }
    catch (std::logic_error& err)
    {
        const string errstr{(string{"Program options error: "}
            += err.what()) += '\n'};
        cerr << errstr;
        exit(EXIT_SUCCESS);
    }


    if (var_map.count(help_option.option()))
    {
        const string logstr{((string{"Usage: miditext [OPTION]... "
            "[MIDIFILE]\nmiditext Version 1.1.6\n"}
            += lexical_cast<string>(desc)) += '\n')
            += "Report bugs to: janzentome@gmail.com\nmiditext home page: "
               "https://github.com/tomejanzen/textmiditools\n"};
        cout << logstr;
        exit(EXIT_SUCCESS);
    }

    if (var_map.count(version_option.option())) [[unlikely]]
    {
        cout << "miditext\nTextMIDITools Version 1.1.6\n"
            "Copyright © 2025 Thomas E. Janzen\n"
            "License GPLv3+: GNU GPL version 3 "
            "or later <https://gnu.org/licenses/gpl.html>\n"
            "This is free software: you are free to "
            "change and redistribute it.\n"
            "There is NO WARRANTY, to the extent permitted by law.\n";
        exit(EXIT_SUCCESS);
    }

    string midi_filename;
    if (var_map.count(midi_in_option.option()))
    {
        midi_filename = var_map[midi_in_option.option()].as<string>();
        if (!exists(midi_filename))
        {
            const string errstr{((midi_in_option.option() + ' ')
                += midi_filename) += " File does not exist.\n"};
            cerr  << errstr;
            exit(EXIT_SUCCESS);
        }
    }
    else
    {
        const string errstr{(string{"You must provide a MIDI input file\n"}
            += lexical_cast<string>(desc)) += '\n'};
        cerr << errstr;
        exit(EXIT_SUCCESS);
    }
    RhythmRational quantum{0};
    if (var_map.count(quantize_option.option())) [[unlikely]]
    {
        string quantum_string{var_map[quantize_option.option()].as<string>()};
        istringstream iss{quantum_string};
        iss >> quantum;
    }
    bool lazy{};
    if (var_map.count(lazy_option.option()))
    {
        lazy = true;
    }

    bool verbose{};
    if (var_map.count(verbose_option.option())) [[unlikely]]
    {
        verbose = true;
    }
    bool answer{};
    if (var_map.count(answer_option.option())) [[unlikely]]
    {
        answer = true;
    }

    string text_filename;
    if (var_map.count(textmidi_out_option.option()))
    {
        text_filename = var_map[textmidi_out_option.option()].as<string>();
    }
    else
    {
        text_filename = midi_filename + ".txt";
        if (verbose)
        {
            const string logstr{(string{"Will write to "}
                += text_filename) += '\n'};
            cout << logstr;
        }
    }

    {
        string dynamics_configuration_file{};
        if (var_map.count(dynamics_configuration_option.option())) [[unlikely]]
        {
            dynamics_configuration_file
                = var_map[dynamics_configuration_option.option()].as<string>();
        }
        midi::dynamics_map = textmidi::read_dynamics_configuration
            (dynamics_configuration_file);
    }

    bool dotted_rhythms{true};
    if (var_map.count(dotted_rhythms_option.option())) [[unlikely]]
    {
        string dotted_rhythms_string{
            var_map[dotted_rhythms_option.option()].as<string>()};
        to_upper(dotted_rhythms_string);
        dotted_rhythms = ("TRUE" == dotted_rhythms_string);
    }

    if (var_map.count(rhythm_expression_option.option())) [[unlikely]]
    {
        string rhythm_expression_string{
            var_map[rhythm_expression_option.option()].as<string>()};
        to_upper(rhythm_expression_string);
        if (midi::rhythm_expression_map.contains(rhythm_expression_string))
        {
            const RhythmExpression rhythm_expression{midi::rhythm_expression_map
                    [rhythm_expression_string]};
            switch (rhythm_expression)
            {
              case RhythmExpression::Rational:
                textmidi::rational::print_rhythm
                    = make_unique<PrintRhythmRational>(dotted_rhythms);
                break;
              case RhythmExpression::SimpleContinuedFraction:
                textmidi::rational::print_rhythm
                    = make_unique<textmidi::rational
                       ::PrintRhythmSimpleContinuedFraction>();
                break;
            }
        }
    }
    else
    {
        if (var_map.count(dotted_rhythms_option.option())) [[unlikely]]
        {
            textmidi::rational::print_rhythm
                = make_unique<PrintRhythmRational>(dotted_rhythms);
        }
    }

    if (answer && exists(midi_filename))
    {
        const string answer_prompt{(string{"Overwrite "}
                += midi_filename) += "?\n"};
        cout << answer_prompt;
        string answerstr{};
        cin >> answerstr;
        if (!(toupper(answerstr[0]) == 'Y'))
        {
            exit(EXIT_SUCCESS);
        }
    }
    if (midi_filename == text_filename)
    {
        cerr << "The text and MIDI filenames are the same!  "
            "You would have overwritten the MIDI file!; Must exit!\n";
        exit(EXIT_SUCCESS);
    }

    midi::MidiStreamVector midivector{};
    try
    {
        ifstream midifilestr;
        midifilestr.open(midi_filename, std::ios_base::binary);
        if (!midifilestr)
        {
            const string errstr{(string{"Open failed for "} += midi_filename)
                += '\n'};
            cerr << errstr;
            return EXIT_FAILURE;
        }
        const auto len{file_size(midi_filename)};
        midivector.resize(len);
        midifilestr.read(reinterpret_cast<char*>(midivector.data()), len);
        midifilestr.close();
    }
    catch (std::ios_base::failure &iosfail)
    {
        const string errstr{string{iosfail.what()} += '\n'};
        cerr << errstr;
        exit(EXIT_SUCCESS);
    }

    ofstream text_filestr(text_filename.c_str());

    if (!text_filestr)
    {
        const string logstr{(string{"can't open "} += text_filename) += '\n'};
        cout << logstr;
    }

    midi::MidiStreamIterator midiiter{midivector.begin()};

    if (midivector.size() < sizeof(midi::MidiHeader))
    {
        cerr << "ERROR File too short for a MIDI Header\n";
        exit(0);
    }
    midi::MidiHeader midi_header{midivector.data()};
    if (verbose)
    {
        const string logstr{string{lexical_cast<string>(midi_header)} += '\n'};
        cout << logstr;
    }
    text_filestr << "FILEHEADER ";
    midiiter += sizeof(midi::MidiHeader);
    const uint32_t ticksperquarter{midi_header.division_};
    uint32_t ticks_per_whole{
        static_cast<uint32_t>(ticksperquarter * midi::QuartersPerWhole)};

    string textmidi_str{};
    textmidi_str.reserve(32);
    (((((textmidi_str += lexical_cast<string>(midi_header.ntrks_)) += ' ')
       += lexical_cast<string>(midi_header.division_)) += ' ')
       += lexical_cast<string>(midi_header.format_)) += '\n';
    text_filestr << textmidi_str;
    if (verbose)
    {
        const string logstr{(string{"FORMAT: "}
            += lexical_cast<string>(midi_header.format_)) += '\n'};
        cout << logstr;
    }
    size_t track_qty{midi_header.ntrks_};

    vector<TrackPositionLength> track_locations{};
    find_tracks(std::views::all(midivector), track_locations, track_qty);
    vector<DelayEvents> midi_delay_event_tracks(track_locations.size());

#if defined(DEBUG_THREADLESS)
    if (verbose)
    {
        cout << "Running unthreaded\n";
    }
    for (int32_t i{}; auto& track_loc : track_locations)
    {
        MidiStreamRange track_stream{midivector.begin() + track_loc.first,
            midivector.begin() + track_loc.first + track_loc.second};
        ConvertTrack convert_track{midi_delay_event_tracks[i], ticks_per_whole, quantum,
            track_stream};
        convert_track();
        ++i;
    }
#else
    if (verbose)
    {
        cout << "Running threaded\n";
    }
    {
        vector<std::jthread> track_threads(track_locations.size());

        for (int32_t i{}; auto& track_loc : track_locations)
        {
            MidiStreamRange track_stream{midivector.begin() + track_loc.first,
                midivector.begin() + track_loc.first + track_loc.second};
            ConvertTrack convert_track(midi_delay_event_tracks[i], ticks_per_whole,
                    quantum, track_stream);
            std::jthread track_thread{convert_track};
            track_threads[i] = std::move(track_thread);
            ++i;
        }
    }
#endif
    if (verbose)
    {
        const string logstr{(string{"Will write to: "} += text_filename)
            += '\n'};
        cout << logstr;

        cout << "Events per track:\n";
        for (int32_t i{}; auto& mdep : midi_delay_event_tracks)
        {
            cout << "  Track: " << setw(3) << (i + 1)
                 << ": " << setw(8) << mdep.size() << '\n';
            ++i;
        }
    }
    int64_t quantized_rhythms_count{};
    int64_t non_quantized_rhythms_count{};
    int64_t zero_rhythms_count{};
    for (auto& mdet : midi_delay_event_tracks)
    {
        text_filestr << "\nSTARTTRACK\n";
        if (verbose)
        {
            // If a score was entered in a scoring program rather
            // than played in,
            // it should largely contain delays that are simply related to
            // whole notes and their musical ratios will have low divisors.
            const set<int64_t> musical_divisors{1, 2, 4, 6, 8, 12, 16};
            const auto ticks_per_whole_local{
                midi::QuartersPerWhole * ticksperquarter};
            zero_rhythms_count  += count_if(mdet, [](const DelayEvent& mde) { return !mde.first; } );
            quantized_rhythms_count += count_if(mdet,
                [ticks_per_whole_local, musical_divisors](const DelayEvent& mde) {
                const RhythmRational rr{mde.first, static_cast<int64_t>(ticks_per_whole_local)};
                return musical_divisors.contains(rr.denominator()); } );
            non_quantized_rhythms_count += count_if(mdet,
                [ticks_per_whole_local, musical_divisors](const DelayEvent& mde) {
                const RhythmRational rr{mde.first,
                    static_cast<int64_t>(ticks_per_whole_local)};
                return !musical_divisors.contains(rr.denominator()); } );
        }
        if (lazy)
        {
            PrintLazyTrack print_lazy_track{mdet, quantum, ticksperquarter};
            text_filestr << print_lazy_track << '\n';
        }
        else
        {
            copy(mdet, std::ostream_iterator<DelayEvent>(text_filestr, "\n"));
        }
    }
    if (verbose)
    {
        quantized_rhythms_count -= zero_rhythms_count;
        const double quantized_rhythms_fraction{
            static_cast<double>(quantized_rhythms_count)
            / static_cast<double>(non_quantized_rhythms_count
            + quantized_rhythms_count)};
        cout << "Quantized rhythms: " << (quantized_rhythms_fraction * 100.0) << "%\n";
    }
    return EXIT_SUCCESS;
}
